sdl2-vga-terminal/src/TuiTerminal.cpp

```diff
@@ -0,0 +1,343 @@
+#include "TuiTerminal.hpp"
+#include <algorithm>
+#include "tui/utils.hpp"
+
+
+TuiTerminal::TuiTerminal(const std::string& title, const std::string& description):
+	_term(VgaTerminal("VgaTerminal TUI mode")), _title(title), _desc(description)
+{
+	_term.autoScroll = false;
+	_term.showCursor = false;
+}
+
+
+void TuiTerminal::draw()
+{
+	_term.resetViewport();
+	_drawBackGround();
+	_drawHeader();
+	_drawFooter();
+	_term.setViewPort(0, 5, _term.getMode().tw, _term.getMode().th - 5);
+}
+
+void TuiTerminal::render()
+{
+	_term.render();
+}
+
+void TuiTerminal::_drawBackGround() noexcept
+{
+	_term.fill(background_tc.c, background_tc.col, background_tc.bgCol);
+}
+
+void TuiTerminal::_drawHeader() noexcept
+{
+	// TODO aligment
+	// split in 3 parts, left, middle, right
+	// 3 rows, columns = 
+	constexpr uint8_t x = 4;
+	constexpr uint8_t y = 0;
+	constexpr uint8_t w = 72;
+	constexpr uint8_t h = 3;
+	constexpr uint8_t col = 14;
+	constexpr uint8_t bgCol = 1;
+	constexpr SDL_Rect r{ x, y, w, h };
+	
+	
+	const std::string version = std::string("Version ") + VERSION;
+	auto [lmark, rmark] = tui::align3Strings(_title, _desc, version, w);
+	const SDL_Rect rl = { x, y, lmark, h };
+	const SDL_Rect rm = { x+lmark-1, y, rmark-lmark+1, h };
+	const SDL_Rect rr = { x+rmark-1, y, w-rmark+1, h };
+
+	_drawLeftBorder(rl, col, bgCol, _title);
+	_drawMiddleBorder(rm, col, bgCol, _desc);
+	_drawRightBorder(rr, col, bgCol, version);
+	_drawShadow(r);
+}
+
+void TuiTerminal::drawSingleBorderDialog(const SDL_Rect& r, uint8_t col, uint8_t bgCol, const std::string& title, const std::string body)
+{
+	if (r.h < 3) {
+		throw std::invalid_argument("height must be 3 at least");
+	}
+
+	if (r.w < title.size() + 2) {
+		throw std::invalid_argument("title too long");
+	}
+
+	if (r.w < body.size() + 2) {
+		throw std::invalid_argument("body too long");
+	}
+
+
+	std::string padTitle = tui::alignString(title, r.w - 2 - title.size(), 196);
+	std::string padBody = tui::alignString(body, r.w - 2 - body.size(), ' ');
+	// Top
+	_term.gotoXY(r.x, r.y);
+	_term.write(218, col, bgCol);
+	_term.write(padTitle, col, bgCol);
+	_term.write(191, col, bgCol);
+	// Middle
+	_term.gotoXY(r.x, r.y + 1);
+	_term.write(179, col, bgCol);
+	_term.write(padBody, col, bgCol);
+	_term.write(179, col, bgCol);
+	for (int i = 2; i < r.h - 1; i++)
+	{
+		_term.gotoXY(r.x, r.y + i);
+		_term.write(179, col, bgCol);
+		for (int j = 2; j < r.w; j++) {
+			_term.write(' ', col, bgCol);
+		}
+		_term.write(179, col, bgCol);
+	}
+	// Bottom
+	_term.gotoXY(r.x, r.y + r.h - 1);
+	_term.write(192, col, bgCol);
+	_drawHBorder(r.x + 1, r.y + r.h - 1, r.w, col, bgCol, 196);
+	_term.write(217, col, bgCol);
+	_drawShadow(r);
+}
+
+void TuiTerminal::drawDialog(const SDL_Rect& r, uint8_t col, uint8_t bgCol, const std::string& header, const std::string body)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Function parameter 'body' should be passed by const reference.](https://app.codacy.com/manual/Raffaello/sdl2-vga-terminal/pullRequest?prid=5472420)